frame/frame.go

```diff
@@ -146,6 +146,58 @@ func (f *Frame) String() string {
 	return fmt.Sprintf("{header: %v, body: %v}", f.Header, f.Body)
 }
 
+// Performs a deep copy of a frame object
+func (f* Frame) Clone() *Frame {
+	return &Frame{
+		Header: f.Header.Clone(),
+		Body:   f.Body.Clone(),
+	}
+}
+
+// Performs a deep copy of a header object and returns the new object.
+func (h *Header) Clone() *Header {
+	newHeader := *h // it's only value types so this is fine
+	return &newHeader
+}
+
+// Performs a deep copy of a body object and returns the new object.
+func (b *Body) Clone() *Body {
+	var tracingId *primitive.UUID
+	if b.TracingId == nil {
```

Comment:
The implementation needs to be different if we use pointer types because the assignment will just copy the pointer value instead of the UUID value. In that case we can test the nilness inside the `Clone()` method but this will cause `panic` if in the future we add an interface for the `UUID` type and we use that interface.

Reply:
Actually my comment is not correct, we can pretty much keep the same implementation.

Reply:
done